### PR TITLE
Extend the instruction set

### PR DESCRIPTION
**Pull request by rkilar**
I renamed some classes to avoid further confusion.

### CODE REVIEW

File: app/src/main/kotlin/compiler/intermediate_form/Instruction.kt [hunk 1]
```diff
@@ -1,61 +1,152 @@
...
+        }
+
+        // Bitwise instructions
+        data class NotR(override val reg: Register) : OneRegIns() //                                NOT  reg          reg  <- ~reg
+        data class AndRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() // AND  reg0, reg1   reg0 <- reg0 & reg1
+        data class OrRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() //  OR   reg0, reg1   reg0 <- reg0 & reg1
+        data class XorRR(override val reg0: Register, override val reg1: Register) : TwoRegIns() // XOR  reg0, reg1   reg0 <- reg0 ^ reg1
+        data class ShiftLeftR(val reg: Register) : InPlaceInstruction() { //                        SAL  reg,  CL     reg  <- reg << CL
+            override val regsDefined: Collection<Register> = setOf(reg)
+            override val regsUsed: Collection<Register> = setOf(Register.RCX)
```
Comment (bebidek) [inline app/src/main/kotlin/compiler/intermediate_form/Instruction.kt:115]: We **use** reg here as well. Same with right shift.

File: app/src/main/kotlin/compiler/intermediate_form/Instruction.kt [hunk 2]
```diff
@@ -1,61 +1,152 @@
...
+        sealed class SetInstruction : InPlaceInstruction() {
+            abstract val reg: Register
+            override val regsDefined: Collection<Register> get() = setOf(reg)
+            // I'm unsure if this is needed
+            // The SETcc instructions set the lower 8 bits of a register
+            // so the value after SETcc is dependent on the value of reg before instruction.
+            // Specifically, if this wasn't included in when doing
+            // 1: XOR    reg,   reg   ; reset to zero
+            // 2: CMP    rega,  regb
+            // 3: SETcc  reg          ; set lower 8 bits depending on comparison
```
Comment (bebidek) [inline app/src/main/kotlin/compiler/intermediate_form/Instruction.kt:140]: Docs say (and GCC confirms that, so it'll probably be the same with NASM) that operation SETcc will not accept 64-bit register at all.

**Review by bebidek — COMMENTED** [review]
Few comments regarding the assembly instructions themselves.

**Review by JHodorHodor — CHANGES_REQUESTED** [review]
Kilka stylistycznych komentarzy, plus rzeczy od Bartka. Poza tym wygląda to bardzo dobrze.

File: app/src/main/kotlin/compiler/intermediate_form/Instruction.kt (hunk 2 above)
Comment (rkilar) [reply app/src/main/kotlin/compiler/intermediate_form/Instruction.kt:140]: Yeah, I know that SETcc takes 8-bit registers. I meant for Setcc(reg) to represent SETcc with low 8-bit register corresponding to reg as an operand. I read in documentation that the upper bits are unchanged for 8-bit operands. I'm not sure if the byte order is correct (so setting AL to 1 is the same as setting RAX to 1). I haven't tested it and I'm not all that versed in assembly.

Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/Instruction.kt:144]: According to AMD64 docs, smaller registers (such as AL, AX, EAX) occupy lower bits of full register, so the byte order should be correct.

File: app/src/main/kotlin/compiler/intermediate_form/IFTPattern.kt [hunk 3]
```diff
@@ -0,0 +1,174 @@
+package compiler.intermediate_form
+
+import kotlin.reflect.KClass
+import kotlin.reflect.safeCast
+
+// Class that represents a pattern of an intermediate form tree
+// Not to be confused with the Pattern interface
+sealed class IFTPattern {
+
+    data class MatchResult(val subtrees: List<IntermediateFormTreeNode>, val args: Map<String, Any>)
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/IFTPattern.kt:10]:
```suggestion
    data class MatchResult(val unmatchedSubtrees: List<IntermediateFormTreeNode>, val matchedValues: Map<String, Any>)
```
maybe?

**Review by JHodorHodor — APPROVED** [review]
Lgtm, dodałem jeszcze jeden malutki komentarz. Po rozwiązaniu kwesti z komentarza Bartka można mergować.

**Pull request merged**